apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/conf/dynamic/ConfigurationDiscoveryService.java

```diff
@@ -0,0 +1,241 @@
+        );
+    }
+
+    @Override
+    public void onComplete() throws Throwable {
+
+    }
+
+    @Override
+    public void shutdown() throws Throwable {
+        if (getDynamicConfigurationFuture != null) {
+            getDynamicConfigurationFuture.cancel(true);
+        }
+    }
+
+    /**
+     * Register dynamic configuration watcher.
+     *
+     * @param watcher dynamic configuration watcher
+     */
+    public void registerAgentConfigChangeWatcher(AgentConfigChangeWatcher watcher) {
+        WatcherHolder holder = new WatcherHolder(watcher);
+        if (register.containsKey(holder.getKey())) {
+            throw new IllegalStateException("Duplicate register, watcher=" + watcher);
+        }
+        register.put(holder.getKey(), holder);
+    }
+
+    /**
+     * Process ConfigurationDiscoveryCommand and notify each configuration watcher.
+     *
+     * @param configurationDiscoveryCommand Describe dynamic configuration information
+     */
+    public void handleConfigurationDiscoveryCommand(ConfigurationDiscoveryCommand configurationDiscoveryCommand) {
+        final String responseUuid = configurationDiscoveryCommand.getUuid();
+        final List<KeyStringValuePair> config = configurationDiscoveryCommand.getConfig();
+
+        if (responseUuid != null && Objects.equals(this.uuid, responseUuid)) {
+            return;
+        }
+
+        config.forEach(property -> {
+            String propertyKey = property.getKey();
+            WatcherHolder holder = register.get(propertyKey);
+            if (holder != null) {
+                AgentConfigChangeWatcher watcher = holder.getWatcher();
+                String newPropertyValue = property.getValue();
+                if (newPropertyValue == null) {
+                    if (watcher.value() != null) {
+                        // Notify watcher, the new value is null with delete event type.
+                        watcher.notify(
+                            new AgentConfigChangeWatcher.ConfigChangeEvent(
+                                null, AgentConfigChangeWatcher.EventType.DELETE
+                            ));
+                    } else {
+                        // Don't need to notify, stay in null.
+                    }
+                } else {
+                    if (!newPropertyValue.equals(watcher.value())) {
+                        watcher.notify(new AgentConfigChangeWatcher.ConfigChangeEvent(
+                            newPropertyValue, AgentConfigChangeWatcher.EventType.MODIFY
+                        ));
+                    } else {
+                        // Don't need to notify, stay in the same config value.
+                    }
+                }
+            } else {
+                LOGGER.warn("Config {} from OAP, doesn't match any watcher, ignore.", propertyKey);
+            }
+        });
+        this.uuid = responseUuid;
+
+        LOGGER.trace("Current configurations after the sync." + LINE_SEPARATOR + register.toString());
```

Comment:
Don't use `+`, this links the string always, no matter what is current log level. 
Use `public void trace(final String message, final Object... objects)` instead.